libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteDriver.kt

```diff
@@ -0,0 +1,118 @@
+package com.mapbox.navigation.core.replay.route2
+
+import com.mapbox.api.directions.v5.DirectionsCriteria
+import com.mapbox.api.directions.v5.models.LegAnnotation
+import com.mapbox.api.directions.v5.models.RouteLeg
+import com.mapbox.geojson.LineString
+import com.mapbox.geojson.Point
+import com.mapbox.geojson.utils.PolylineUtils
+import com.mapbox.turf.TurfConstants
+import com.mapbox.turf.TurfMeasurement
+
+internal class ReplayRouteDriver {
+
+    private val routeSmoother = ReplayRouteSmoother()
+    private val routeInterpolator = ReplayRouteInterpolator()
+
+    private var timeMillis = 0L
+
+    /**
+     * Given a geometry polyline, simulate the locations needed to drive the geometry.
+     *
+     * @param options allow you to control the driver and car behavior
+     * @param geometry polyline string at precision 6
+     */
+    fun driveGeometry(options: ReplayRouteOptions, geometry: String): List<ReplayRouteLocation> {
+        val coordinates = LineString.fromPolyline(geometry, 6).coordinates()
+        return drivePointList(options, coordinates)
+    }
+
+    /**
+     * Given a list of location points, simulate the locations needed to drive the points.
+     *
+     * @param options allow you to control the driver and car behavior
+     * @param points list of points describing a route
+     */
+    fun drivePointList(options: ReplayRouteOptions, points: List<Point>): List<ReplayRouteLocation> {
+        val distinctPoints = routeSmoother.distinctPoints(points, ReplayRouteSmoother.DISTINCT_POINT_METERS)
+        if (distinctPoints.size < 2) return emptyList()
+
+        val smoothLocations = routeInterpolator.createSpeedProfile(options, distinctPoints)
+        val replayRouteLocations = mutableListOf<ReplayRouteLocation>()
+        replayRouteLocations.addLocation(smoothLocations.first())
+
+        for (i in 0 until smoothLocations.lastIndex) {
+            val segmentStart = smoothLocations[i]
+            val segmentEnd = smoothLocations[i + 1]
+            val segmentRoute = routeSmoother.segmentRoute(
+                distinctPoints,
+                segmentStart.routeIndex!!,
+                segmentEnd.routeIndex!!)
+            replayRouteLocations.addInterpolatedLocations(options, segmentRoute, segmentStart, segmentEnd)
+        }
```

Comment:
What about extracting this `for` loop into a private function and give them a proper name?

Reply:
yeah iterated some more =)